File: test/vlasov_poisson.py
import numpy as np
from tool_box import integrate
from equilibrium_manager import EquilibriumManager
from scheme import Scheme
from output_manager import OutputManager


def perturbate_func(mesh_x, mesh_v, f, p):
    x = mesh_x.x
    v = mesh_v.x
    nx = mesh_x.nx
    nv = mesh_v.nx
    return np.array([[(1 + p(x[i], v[j])) * f[i, j]
                     for j in range(nv)] for i in range(nx)])


def get_equilibriums(eq_manager, perturbated=False, epsilon=1e-3):
    mesh_x = eq_manager.mesh_x
    mesh_v = eq_manager.mesh_v
    x_min = mesh_x.x_min
    x_max = mesh_x.x_max

    if perturbated:
        L = 2 * np.pi / (x_max - x_min)
        p = lambda xi, vj: epsilon * (np.cos(L * xi + 1.) + 0.5 * np.sin(2 * L * xi - 0.5))
        dx_p = lambda xi, vj: epsilon * (-L * np.sin(L * xi + 1.) + 0.5 * 2 * L * np.cos(2 * L * xi - 0.5)) - 1

        fe_eq = perturbate_func(mesh_x, mesh_v, eq_manager.fe_eq, p)
        fi_eq = perturbate_func(mesh_x, mesh_v, eq_manager.fi_eq, p)
        dv_fe_eq = perturbate_func(mesh_x, mesh_v, eq_manager.dv_fe_eq, p)
        dv_fi_eq = perturbate_func(mesh_x, mesh_v, eq_manager.dv_fi_eq, p)
        dx_fe_eq = (perturbate_func(mesh_x, mesh_v, eq_manager.dx_fe_eq, p)
            + perturbate_func(mesh_x, mesh_v, eq_manager.fe_eq, dx_p))
        dx_fi_eq = (perturbate_func(mesh_x, mesh_v, eq_manager.dx_fi_eq, p)
            + perturbate_func(mesh_x, mesh_v, eq_manager.fi_eq, dx_p))

        scale_coef = integrate(mesh_x, mesh_v, fi_eq) / integrate(mesh_x, mesh_v, fe_eq)
        fe_eq *= scale_coef
        dv_fe_eq *= scale_coef
        dx_fe_eq *= scale_coef
    else:
        fe_eq = eq_manager.fe_eq
        fi_eq = eq_manager.fi_eq
        dv_fe_eq = eq_manager.dv_fe_eq
        dv_fi_eq = eq_manager.dv_fi_eq
        dx_fe_eq = eq_manager.dx_fe_eq
        dx_fi_eq = eq_manager.dx_fi_eq

    return fe_eq, fi_eq, dx_fe_eq, dx_fi_eq, dv_fe_eq, dv_fi_eq


def vlasov_poisson(data, mesh_x, mesh_v):
    tf = data.T_final
    nt = data.nb_time_steps
    dt = tf / nt

    eq_manager = EquilibriumManager(mesh_x, mesh_v, data.coef)
    output = OutputManager(data, mesh_x, mesh_v, eq_manager.fe_eq, eq_manager.fi_eq)

    fe_eq, fi_eq, dx_fe_eq, dx_fi_eq, dv_fe_eq, dv_fi_eq = get_equilibriums(
        eq_manager, perturbated=False, epsilon=1e-2)

    fe = perturbate_func(mesh_x, mesh_v, eq_manager.fe_eq,
        data.perturbation_init)
    fi = perturbate_func(mesh_x, mesh_v, eq_manager.fi_eq,
        data.perturbation_init)

    scheme = Scheme(mesh_x, mesh_v, fe, fi, fe_eq, fi_eq,
    dx_fe_eq, dx_fi_eq, dv_fe_eq, dv_fi_eq, data.wb_scheme)

    for it in range(0, nt):
        if data.output and it % data.freq_output == 0:
            print("Computing time ", it * dt)

        if data.wb_scheme and data.projection and (it % data.freq_projection == 0):
            if data.output and it % data.freq_output == 0 :
                print("Projecting...")

            scheme.project(eq_manager, data.projection_type)

        if it % data.freq_save == 0:
            if data.output:
                print("Saving solution...")
                output.save_output(scheme.fe, scheme.fi, scheme.fe_eq,
                    scheme.fi_eq, it * dt)

        scheme.compute_iteration(dt)

    output.end_of_simulation()

    if data.output:
        print("-" * 40)
        print(" Computation ends at T = ", data.T_final)
        print("-" * 40)

File: examples/vlasov_poisson.jl
using VlasovPoissonTwoSpecies
using Test
using FFTW
using Plots
import SpecialFunctions: ellipk
using ProgressMeter

refine_dt              = 1
refine_factor          = 1
eps_perturbation       = 0.001

coef                   = Coef()
data                   = Data()
data.wb_scheme         = true
data.projection        = false
data.projection_type   = :coefficients   # :BGK
data.T_final           = 200
data.nb_time_steps     = 1000 * refine_factor * refine_dt
data.nx                = 64 * refine_factor
data.nv                = 64 * refine_factor
data.x_min             = 0
data.x_max             = 4 * ellipk(coef.m)
data.v_min             = -10
data.v_max             = 10
data.perturbation_init = (x, v) -> eps_perturbation * (cos(2π * x / (data.x_max - data.x_min) + 1))
data.output            = true
data.vtk               = true
data.freq_save         = 5 * refine_factor * refine_dt
data.freq_output       = 5 * refine_factor * refine_dt
data.freq_projection   = 5 * 5 * refine_factor * refine_dt

function run(coef::Coef, data::Data)

    mesh_x = Mesh(data.x_min, data.x_max, data.nx)
    mesh_v = Mesh(data.v_min, data.v_max, data.nv)
     
    tf = data.T_final
    nt = data.nb_time_steps
    dt = tf / nt
    
    eq_manager = EquilibriumManager(coef, mesh_x, mesh_v)
    output = OutputManager(data, mesh_x, mesh_v, eq_manager.fe, eq_manager.fi)
    
    fe_eq, fi_eq, dx_fe_eq, dx_fi_eq, dv_fe_eq, dv_fi_eq =
        get_equilibriums(eq_manager, perturbated = false, epsilon = 1e-2)
    
    fe = perturbate_func(mesh_x, mesh_v, eq_manager.fe, data.perturbation_init)
    fi = perturbate_func(mesh_x, mesh_v, eq_manager.fi, data.perturbation_init)
    
    scheme = Scheme( mesh_x, mesh_v, fe, fi, fe_eq, fi_eq, dx_fe_eq, dx_fi_eq,
        dv_fe_eq, dv_fi_eq, data.wb_scheme,)
    
    v = mesh_v.x
    
    e_fe, e_fi = compute_normalized_energy(output, fe, fi)
    
    @showprogress 1 for i in 1:nt
    
        compute_iteration(scheme, dt)
        
        save(output, scheme, i*dt)
    
    end

    output.t, output.energy_fe, output.energy_fi

end

time, energy_fe, energy_fi = run(coef, data)

plot(time, energy_fe)
plot!(time, energy_fi)

